File: app/desktop/lib/parser.js
// Pure CommonJS markdown parser for the Vibe Assembly .ai/ workspace state.
// No dependencies — runs identically in Electron main, tests, and Node.
'use strict';

function between(text, startMarker, endMarker) {
  const i = text.indexOf(startMarker);
  if (i === -1) return undefined;
  const rest = text.slice(i + startMarker.length);
  if (endMarker == null) return rest.trim();
  const j = rest.indexOf(endMarker);
  return (j === -1 ? rest : rest.slice(0, j)).trim();
}

/** Parse `- **field:** value` lines anywhere in a section (case-insensitive). */
function field(text, name) {
  const m = text.match(new RegExp(`^- \\*\\*${name}:\\*\\*\\s*(.*)$`, 'mi'));
  return m ? m[1].trim() : undefined;
}

// Normalize CRLF (Windows checkouts via git autocrlf) so field captures
// never carry a trailing \r — otherwise status comparisons fail silently.
function normalize(raw) {
  return String(raw ?? '').replace(/\r\n?/g, '\n');
}

/**
 * Parse `.ai/active_task.md` per the ticket contract:
 * Wave, Status Snapshot, Active Ticket pointer, Queue table, per-ticket sections.
 */
function parseActiveTask(raw) {
  const text = normalize(raw);
  const queueSection = between(text, '## Queue', '\n## ') || '';

  const rows = [];
  for (const line of queueSection.split('\n')) {
    const m = line.match(/^\|\s*(\d+)\s*\|\s*(T-\d+)\s*\|(.*?)\|(.*?)\|(.*?)\|(.*?)\|/);
    if (!m) continue;
    const cells = m.slice(3).map((c) => c.trim());
    rows.push({
      order: Number(m[1]),
      id: m[2],
      title: cells[0],
      status: cells[1] || '',
      effort: cells[2] || '',
      depends_on: cells[3] || '',
    });
  }

  const pointerMatch = text.match(/^##\s*Active Ticket\s*[\r\n]+- \*\*id:\*\*\s*(.+)$/m);
  const activeId = pointerMatch ? pointerMatch[1].trim() : undefined;

  // Per-ticket detail section: "## Ticket T-NNN" up to the next "## " heading.
  let active = null;
  if (activeId && activeId !== 'none') {
    const sec = between(text, `## Ticket ${activeId}`, '\n## ');
    if (sec) {
      const relayIdx = sec.indexOf('### relay_notes');
      // Allowlist lives under "### Meta" only — never absorb Steps bullets.
      const metaSec = between(sec, '### Meta', '\n### ') || '';
      active = {
        id: activeId,
        title: queueRows(rows, activeId).title,
        status: field(metaSec, 'status'),
        effort: field(metaSec, 'effort'),
        files_allowed: [...metaSec.matchAll(/^\s+- (.+)$/gm)]
          .map((m) => m[1].trim())
          .filter((p) => p !== 'none' && !p.startsWith('(')), // drop "(default: ...)"
        context_inline: between(sec, '### context_inline', '\n### '),
        steps: parseSteps(between(sec, '### Steps', '\n### ')),
        done_when: bullets(between(sec, '### done_when', '\n### ')),
        stop_when: bullets(between(sec, '### stop_when', '\n### ')),
        relay_notes: bullets(relayIdx === -1 ? '' : sec.slice(relayIdx + '### relay_notes'.length)),
      };
    }
  }

  const snapshot = {
    current_ticket: field(between(text, '## Status Snapshot', '\n## ') || '', 'current_ticket'),
    completed_tickets: field(between(text, '## Status Snapshot', '\n## ') || '', 'completed_tickets'),
    remaining_tickets: field(between(text, '## Status Snapshot', '\n## ') || '', 'remaining_tickets'),
  };

  const waveName = field(between(text, '## Wave', '\n## ') || '', 'name');

  return { wave_name: waveName, snapshot, active_id: activeId || null, queue: rows, active };
}

function queueRows(rows, id) {
  return rows.find((r) => r.id === id) || {};
}

function parseSteps(section) {
  if (!section) return [];
  const steps = [];
  let cur = null;
  for (const line of section.split('\n')) {
    const step = line.match(/^\s*(\d+)\.\s+(.*)$/);
    if (step) {
      if (cur) steps.push(cur);
      cur = { n: Number(step[1]), text: step[2].trim(), verify: null };
      continue;
    }
    const verify = line.match(/\*\*Verify:\*\*\s*`([^`]+)`\s*(?:→|->)?\s*(.*)/);
    if (verify && cur) cur.verify = { command: verify[1].trim(), expected: (verify[2] || '').trim() };
  }
  if (cur) steps.push(cur);
  return steps;
}

function bullets(section) {
  if (!section) return [];
  return section
    .split('\n')
    .map((l) => l.trim())
    .filter((l) => l.startsWith('- '))
    .map((l) => l.replace(/^- /, '').replace(/\s*\(empty until Code runs\)\s*/i, '').trim())
    .filter((l) => l.length > 0);
}

/**
 * Parse the simple bullet-dash state file (`# Title` + `- **k:** v`).
 * Duplicate keys (e.g. several KB entries each with "Date:") are collected
 * into an array instead of silently overwriting earlier values.
 */
function parseStateFile(raw) {
  const text = normalize(raw);
  const title = (text.match(/^#\s+(.+)$/m) || [, ''])[1].trim();
  const fields = {};
  for (const line of text.split('\n')) {
    const m = line.match(/^- \*\*(.+?):\*\*\s*(.*)$/);
    if (!m) continue;
    const key = m[1].trim();
    const value = m[2].trim();
    if (key in fields) {
      if (Array.isArray(fields[key])) fields[key].push(value);
      else fields[key] = [fields[key], value];
    } else {
      fields[key] = value;
    }
  }
  return { title, fields };
}

/**
 * Fail-closed consistency checks mirroring .ai/ticket_contract.md hard rules.
 * Returns a human-readable violation string, or null when state is coherent.
 */
function validateContract(parsed) {
  const q = parsed.queue || [];
  const id = parsed.active_id;

  // Duplicate ids make the pointer ambiguous even before other checks.
  const seen = new Set();
  for (const r of q) {
    if (seen.has(r.id)) return `Queue contains duplicate id ${r.id}.`;
    seen.add(r.id);
  }

  const workRemains = q.some((r) => r.status !== 'done');

  if (!id) return 'Active Ticket pointer section is missing.';
  if (id === 'none') {
    // Contract: "none" is invalid while any Queue row is not done.
    if (workRemains) {
      const open = q.filter((r) => r.status !== 'done').map((r) => r.id).join(', ');
      return `Active Ticket is "none" but unfinished Queue rows remain: ${open}.`;
    }
    return null;
  }

  const pointed = q.filter((r) => r.id === id);
  if (pointed.length !== 1) {
    return `Queue must contain exactly one row for Active Ticket ${id}, found ${pointed.length}.`;
  }
  const row = pointed[0];
  if (!['pending', 'in_progress'].includes(row.status)) {
    return `Active Ticket ${id} has Queue status "${row.status}" — expected pending/in_progress.`;
  }
  if (parsed.active?.status && parsed.active.status !== row.status) {
    return `Status mismatch: Meta "${parsed.active.status}" vs Queue "${row.status}" for ${id}.`;
  }

  // Every dependency must exist and be done.
  const deps = String(row.depends_on || '')
    .split(',')
    .map((d) => d.trim())
    .filter((d) => d && d.toLowerCase() !== 'none');
  for (const dep of deps) {
    const depRow = q.find((r) => r.id === dep);
    if (!depRow) return `${id} depends_on ${dep}, which is not in the Queue.`;
    if (depRow.status !== 'done') {
      return `${id} depends_on ${dep}, whose status is "${depRow.status}" — must be done.`;
    }
  }

  return null;
}

/**
 * Relay-evidence semantics shared by Mission Control and `va`.
 * A step counts as verified ONLY when a SINGLE relay note both:
 *   - references that exact step number (word-boundary matched, so
 *     "Step 12" never satisfies "Step 1"), and
 *   - carries success evidence (PASS / exit code 0) within that same note.
 * Cross-note leakage (one PASS verifying every step) is impossible here.
 */
function verifiedSteps(steps, relayNotes) {
  const notes = (relayNotes || []).map((n) => String(n).toLowerCase());
  const verified = new Set();
  (steps || []).forEach((_, idx) => {
    const n = idx + 1;
    // \b keeps "step 12" from matching "step 1" (prefix trap).
    const mentionsThisStep = new RegExp(`\\bstep\\s*${n}\\b`);
    const successEvidence = /\bpass\b|\bexit(\s*code)?\s*[:=]?\s*0\b/;
    if (notes.some((note) => mentionsThisStep.test(note) && successEvidence.test(note))) {
      verified.add(n);
    }
  });
  return [...verified];
}

module.exports = { parseActiveTask, parseStateFile, parseSteps, bullets, validateContract, verifiedSteps };

File: app/desktop-rust/src/parser.rs
//! Port of `app/desktop/lib/parser.js` to Rust.
//!
//! The contract semantics are load-bearing: `verified_steps` decides where
//! Code resumes, and `validate_contract` fails closed. Both are ported
//! literally rather than "improved" so the native app and the `va` CLI can
//! never disagree about what state a workspace is in.

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct QueueRow {
    pub order: i64,
    pub id: String,
    pub title: String,
    pub status: String,
    pub effort: String,
    pub depends_on: String,
}

impl QueueRow {
    pub fn is_done(&self) -> bool {
        self.status == "done"
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Step {
    pub n: i64,
    pub text: String,
    pub verify_command: Option<String>,
    pub verify_expected: Option<String>,
}

#[derive(Debug, Clone, Default)]
pub struct Ticket {
    pub id: String,
    pub title: String,
    pub status: Option<String>,
    pub effort: Option<String>,
    pub files_allowed: Vec<String>,
    pub context_inline: String,
    pub steps: Vec<Step>,
    pub done_when: Vec<String>,
    pub stop_when: Vec<String>,
    pub relay_notes: Vec<String>,
}

#[derive(Debug, Clone, Default)]
pub struct Snapshot {
    pub current_ticket: Option<String>,
    pub completed_tickets: Option<String>,
    pub remaining_tickets: Option<String>,
}

#[derive(Debug, Clone, Default)]
pub struct ActiveTask {
    pub wave_name: Option<String>,
    pub snapshot: Snapshot,
    pub active_id: Option<String>,
    pub queue: Vec<QueueRow>,
    pub active: Option<Ticket>,
}

impl ActiveTask {
    /// The resume point: the first step without durable verification evidence.
    /// This is the single most decision-relevant number in the whole dataset.
    pub fn verified_steps(&self) -> Vec<i64> {
        match &self.active {
            Some(t) => verified_steps(&t.steps, &t.relay_notes),
            None => Vec::new(),
        }
    }

    pub fn resume_step(&self) -> Option<i64> {
        let verified = self.verified_steps();
        let total = self.active.as_ref()?.steps.len() as i64;
        (1..=total).find(|n| !verified.contains(n))
    }

    pub fn steps_total(&self) -> usize {
        self.active.as_ref().map_or(0, |t| t.steps.len())
    }

    pub fn has_work(&self) -> bool {
        self.queue.iter().any(|r| !r.is_done())
    }
}

/// Text between `start_marker` and the first `end_marker` after it.
fn between<'a>(text: &'a str, start_marker: &str, end_marker: Option<&str>) -> Option<&'a str> {
    let i = text.find(start_marker)?;
    let rest = &text[i + start_marker.len()..];
    match end_marker {
        None => Some(rest.trim()),
        Some(end) => match rest.find(end) {
            Some(j) => Some(rest[..j].trim()),
            None => Some(rest.trim()),
        },
    }
}

/// Parse a `- **field:** value` line, case-insensitively.
fn field(text: &str, name: &str) -> Option<String> {
    for line in text.lines() {
        let line = line.trim_start();
        let rest = line.strip_prefix("- ")?;
        let rest = rest.strip_prefix("**")?;
        // Split at the first `:**` so field names may contain ':'.
        let end = rest.find(":**")?;
        if !rest[..end].eq_ignore_ascii_case(name) {
            continue;
        }
        return Some(rest[end + 3..].trim().to_string());
    }
    None
}

/// Windows checkouts produce CRLF; strip it so captures never carry a
/// trailing `\r` and silently fail status comparisons.
fn normalize(raw: &str) -> String {
    raw.replace("\r\n", "\n").replace('\r', "\n")
}

fn parse_queue_rows(section: &str) -> Vec<QueueRow> {
    let mut rows = Vec::new();
    for line in section.lines() {
        let line = line.trim();
        if !line.starts_with('|') {
            continue;
        }
        // | order | id | title | status | effort | depends_on |
        let cells: Vec<&str> = line
            .trim_matches('|')
            .split('|')
            .map(str::trim)
            .collect();
        if cells.len() < 6 {
            continue;
        }
        let order: i64 = match cells[0].parse() {
            Ok(n) => n,
            Err(_) => continue,
        };
        if !cells[1].starts_with("T-") {
            continue;
        }
        rows.push(QueueRow {
            order,
            id: cells[1].to_string(),
            title: cells[2].to_string(),
            status: cells[3].to_string(),
            effort: cells[4].to_string(),
            depends_on: cells[5].to_string(),
        });
    }
    rows
}

fn parse_steps(section: &str) -> Vec<Step> {
    let mut steps: Vec<Step> = Vec::new();
    for line in section.lines() {
        let trimmed = line.trim_start();
        // A step is `N. text`.
        if let Some(dot) = trimmed.find(". ") {
            let (head, tail) = trimmed.split_at(dot);
            if let Ok(n) = head.trim().parse::<i64>() {
                if n > 0 && head.trim().chars().all(|c| c.is_ascii_digit()) {
                    steps.push(Step {
                        n,
                        text: tail[2..].trim().to_string(),
                        verify_command: None,
                        verify_expected: None,
                    });
                    continue;
                }
            }
        }
        // `  - **Verify:** \`cmd\` -> expected`
        if let Some(cur) = steps.last_mut() {
            if let Some((cmd, expected)) = parse_verify(trimmed) {
                cur.verify_command = Some(cmd);
                cur.verify_expected = Some(expected);
            }
        }
    }
    steps
}

fn parse_verify(line: &str) -> Option<(String, String)> {
    let after = line.strip_prefix("- **Verify:**")?.trim_start();
    // The command is the text between the first pair of backticks.
    let inner = after.strip_prefix('`')?;
    let end = inner.find('`')?;
    let cmd = inner[..end].to_string();
    let rest = inner[end + 1..].trim_start();
    // Tolerate both the ASCII `->` and the Unicode `→` separator.
    let rest = rest
        .trim_start_matches("->")
        .trim_start_matches('\u{2192}')
        .trim();
    Some((cmd, rest.to_string()))
}

fn bullets(section: &str) -> Vec<String> {
    section
        .lines()
        .map(str::trim)
        .filter(|l| l.starts_with("- "))
        .map(|l| l.trim_start_matches("- ").trim())
        .map(|l| {
            // The contract tells authors to leave an empty relay_notes marker.
            let lower = l.to_ascii_lowercase();
            match lower.find("(empty until code runs)") {
                Some(i) => format!("{}{}", &l[..i], &l[i + "(empty until code runs)".len()..])
                    .trim()
                    .to_string(),
                None => l.to_string(),
            }
        })
        .filter(|l| !l.is_empty())
        .collect()
}

pub fn parse_active_task(raw: &str) -> ActiveTask {
    let text = normalize(raw);
    let queue_section = between(&text, "## Queue", Some("\n## ")).unwrap_or("");
    let queue = parse_queue_rows(queue_section);

    // The Active Ticket pointer is the sole execution pointer. The literal
    // value "none" is meaningful (no work assigned) and must survive here —
    // dropping it would turn a quiescent workspace into a false violation.
    let active_id = text
        .split_once("## Active Ticket")
        .and_then(|(_, rest)| {
            rest.lines()
                .map(str::trim)
                .find_map(|l| l.strip_prefix("- **id:**"))
                .map(|v| v.trim().to_string())
        })
        .filter(|v| !v.is_empty());

    let mut active = None;
    if let Some(id) = active_id.as_deref() {
        if id != "none" {
            if let Some(sec) = between(&text, &format!("## Ticket {id}"), Some("\n## ")) {
                let meta = between(sec, "### Meta", Some("\n### ")).unwrap_or("");
                let title = queue
                    .iter()
                    .find(|r| r.id == id)
                    .map(|r| r.title.clone())
                    .unwrap_or_default();

                // files_allowed is a multi-line list under its own Meta key.
                // Collect only the indented bullets that follow that key, so
                // sibling keys (status/effort/depends_on) are never absorbed.
                let files_allowed = {
                    let mut out = Vec::new();
                    let mut in_block = false;
                    for line in meta.lines() {
                        let t = line.trim();
                        // The key line itself opens the block, then continues.
                        if t.starts_with("- **files_allowed:**") {
                            in_block = true;
                            continue;
                        }
                        if !t.starts_with("- ") {
                            continue;
                        }
                        // Any other `- **key:**` line closes it.
                        if t.starts_with("- **") {
                            in_block = false;
                            continue;
                        }
                        if in_block {
                            if let Some(v) = t.strip_prefix("- ") {
                                let v = v.trim();
                                if v != "none" && !v.starts_with('(') && !v.is_empty() {
                                    out.push(v.to_string());
                                }
                            }
                        }
                    }
                    out
                };

                let relay = sec.find("### relay_notes").map(|i| &sec[i..]);

                active = Some(Ticket {
                    id: id.to_string(),
                    title,
                    status: field(meta, "status"),
                    effort: field(meta, "effort"),
                    files_allowed,
                    context_inline: between(sec, "### context_inline", Some("\n### "))
                        .unwrap_or("")
                        .to_string(),
                    steps: parse_steps(between(sec, "### Steps", Some("\n### ")).unwrap_or("")),
                    done_when: bullets(between(sec, "### done_when", Some("\n### ")).unwrap_or("")),
                    stop_when: bullets(between(sec, "### stop_when", Some("\n### ")).unwrap_or("")),
                    relay_notes: bullets(relay.unwrap_or("")),
                });
            }
        }
    }

    let snap_section = between(&text, "## Status Snapshot", Some("\n## ")).unwrap_or("");
    let snapshot = Snapshot {
        current_ticket: field(snap_section, "current_ticket"),
        completed_tickets: field(snap_section, "completed_tickets"),
        remaining_tickets: field(snap_section, "remaining_tickets"),
    };
    let wave_name = field(
        between(&text, "## Wave", Some("\n## ")).unwrap_or(""),
        "name",
    );

    ActiveTask {
        wave_name,
        snapshot,
        active_id,
        queue,
        active,
    }
}

/// A simple `# Title` + `- **k:** v` file. Duplicate keys become a Vec so
/// per-entry fields (each KB's `Date:`) do not overwrite one another.
#[derive(Debug, Clone, Default)]
pub struct StateFile {
    pub title: String,
    pub fields: Vec<(String, String)>,
}

impl StateFile {
    pub fn get(&self, key: &str) -> Option<&str> {
        self.fields
            .iter()
            .find(|(k, _)| k.eq_ignore_ascii_case(key))
            .map(|(_, v)| v.as_str())
    }

    pub fn all(&self, key: &str) -> Vec<&str> {
        self.fields
            .iter()
            .filter(|(k, _)| k.eq_ignore_ascii_case(key))
            .map(|(_, v)| v.as_str())
            .collect()
    }
}

pub fn parse_state_file(raw: &str) -> StateFile {
    let text = normalize(raw);
    let title = text
        .lines()
        .find_map(|l| l.strip_prefix("# "))
        .unwrap_or("")
        .trim()
        .to_string();
    let mut fields = Vec::new();
    for line in text.lines() {
        let t = line.trim_start();
        let rest = match t.strip_prefix("- **") {
            Some(r) => r,
            None => continue,
        };
        let end = match rest.find(":**") {
            Some(i) => i,
            None => continue,
        };
        fields.push((rest[..end].trim().to_string(), rest[end + 3..].trim().to_string()));
    }
    StateFile { title, fields }
}

/// Split a `## KB-NNN — title` log into discrete entries. The old Electron
/// UI rendered one entry's field labels as if they were separate bugs, so the
/// native app parses entries properly.
#[derive(Debug, Clone)]
pub struct LogEntry {
    pub id: String,
    pub title: String,
    pub body: String,
}

pub fn parse_log_entries(raw: &str) -> Vec<LogEntry> {
    let text = normalize(raw);
    let mut out: Vec<LogEntry> = Vec::new();
    for line in text.lines() {
        if let Some(heading) = line.strip_prefix("## ") {
            let heading = heading.trim();
            // Expected shape: `KB-001 — title` or `[ADR-008] title`.
            let (id, title) = if let Some(rest) = heading.strip_prefix('[') {
                match rest.split_once(']') {
                    Some((id, title)) => (id.trim().to_string(), title.trim().to_string()),
                    None => (String::new(), heading.to_string()),
                }
            } else {
                match heading.split_once("—").or_else(|| heading.split_once(" - ")) {
                    Some((id, title)) => (id.trim().to_string(), title.trim().to_string()),
                    None => (String::new(), heading.to_string()),
                }
            };
            out.push(LogEntry {
                id,
                title,
                body: String::new(),
            });
        } else if let Some(last) = out.last_mut() {
            last.body.push_str(line);
            last.body.push('\n');
        }
    }
    out
}

/// Fail-closed consistency checks mirroring `.ai/ticket_contract.md`.
/// Returns a human-readable violation, or `None` when state is coherent.
pub fn validate_contract(parsed: &ActiveTask) -> Option<String> {
    let q = &parsed.queue;
    let id = parsed.active_id.as_deref();

    // Duplicate ids make the pointer ambiguous before any other check.
    for (i, r) in q.iter().enumerate() {
        if q.iter().skip(i + 1).any(|o| o.id == r.id) {
            return Some(format!("Queue contains duplicate id {}.", r.id));
        }
    }

    let work_remains = q.iter().any(|r| !r.is_done());

    let id = match id {
        None => return Some("Active Ticket pointer section is missing.".to_string()),
        Some("none") => {
            return if work_remains {
                let open = q
                    .iter()
                    .filter(|r| !r.is_done())
                    .map(|r| r.id.as_str())
                    .collect::<Vec<_>>()
                    .join(", ");
                Some(format!(
                    "Active Ticket is \"none\" but unfinished Queue rows remain: {open}."
                ))
            } else {
                None
            }
        }
        Some(other) => other,
    };

    let pointed: Vec<&QueueRow> = q.iter().filter(|r| r.id == id).collect();
    if pointed.len() != 1 {
        return Some(format!(
            "Queue must contain exactly one row for Active Ticket {id}, found {}.",
            pointed.len()
        ));
    }
    let row = pointed[0];
    if row.status != "pending" && row.status != "in_progress" {
        return Some(format!(
            "Active Ticket {id} has Queue status \"{}\" — expected pending/in_progress.",
            row.status
        ));
    }
    if let Some(ticket) = &parsed.active {
        if let Some(status) = &ticket.status {
            if status != &row.status {
                return Some(format!(
                    "Status mismatch: Meta \"{status}\" vs Queue \"{}\" for {id}.",
                    row.status
                ));
            }
        }
    }

    // Every dependency must exist and be done.
    for dep in row
        .depends_on
        .split(',')
        .map(str::trim)
        .filter(|d| !d.is_empty() && !d.eq_ignore_ascii_case("none"))
    {
        match q.iter().find(|r| r.id == dep) {
            None => return Some(format!("{id} depends_on {dep}, which is not in the Queue.")),
            Some(dep_row) if dep_row.status != "done" => {
                return Some(format!(
                    "{id} depends_on {dep}, whose status is \"{}\" — must be done.",
                    dep_row.status
                ))
            }
            Some(_) => {}
        }
    }

    None
}

/// A step counts as verified ONLY when a SINGLE relay note both references
/// that exact step (word-boundary matched, so "step 12" never satisfies
/// "step 1") and carries success evidence. Cross-note leakage is impossible.
pub fn verified_steps(steps: &[Step], relay_notes: &[String]) -> Vec<i64> {
    let lowered: Vec<String> = relay_notes.iter().map(|n| n.to_lowercase()).collect();
    steps
        .iter()
        .filter(|s| {
            lowered.iter().any(|note| {
                mentions_step(note, s.n) && has_success_evidence(note)
            })
        })
        .map(|s| s.n)
        .collect()
}

fn mentions_step(note: &str, n: i64) -> bool {
    let needle = format!("step {n}");
    let hay: Vec<char> = note.chars().collect();
    let nd: Vec<char> = needle.chars().collect();
    if nd.is_empty() || hay.len() < nd.len() {
        return false;
    }
    for start in 0..=(hay.len() - nd.len()) {
        if hay[start..start + nd.len()] != nd[..] {
            continue;
        }
        // \b — a non-word char before, and a non-digit after, so the prefix
        // trap ("step 1" matching inside "step 12") cannot fire.
        let before_ok = start == 0 || !is_word_char(hay[start - 1]);
        let after_idx = start + nd.len();
        let after_ok = after_idx >= hay.len() || !hay[after_idx].is_ascii_digit();
        if before_ok && after_ok {
            return true;
        }
    }
    false
}

fn is_word_char(c: char) -> bool {
    c.is_ascii_alphanumeric() || c == '_'
}

fn has_success_evidence(note: &str) -> bool {
    // \bpass\b
    if word_occurrences(note, "pass") {
        return true;
    }
    // \bexit(\s*code)?\s*[:=]?\s*0\b
    let bytes: Vec<char> = note.chars().collect();
    for i in 0..bytes.len() {
        if !bytes[i..].starts_with(&['e', 'x', 'i', 't']) {
            continue;
        }
        if i > 0 && is_word_char(bytes[i - 1]) {
            continue;
        }
        let mut j = i + 4;
        let mut saw_code = false;
        // optional whitespace then "code"
        let ws: String = bytes[j..].iter().take_while(|c| c.is_whitespace()).collect();
        let rest_start = j + ws.chars().count();
        if note[byte_index(&bytes, rest_start)..].starts_with("code") {
            saw_code = true;
            j = rest_start + 4;
        }
        let _ = saw_code;
        // optional whitespace, optional :/=, optional whitespace, then '0'
        while j < bytes.len() && bytes[j].is_whitespace() {
            j += 1;
        }
        if j < bytes.len() && (bytes[j] == ':' || bytes[j] == '=') {
            j += 1;
        }
        while j < bytes.len() && bytes[j].is_whitespace() {
            j += 1;
        }
        if j < bytes.len() && bytes[j] == '0' {
            let after_ok = j + 1 >= bytes.len() || !is_word_char(bytes[j + 1]);
            if after_ok {
                return true;
            }
        }
    }
    false
}

fn byte_index(chars: &[char], char_idx: usize) -> usize {
    chars[..char_idx].iter().map(|c| c.len_utf8()).sum()
}

fn word_occurrences(hay: &str, needle: &str) -> bool {
    let chars: Vec<char> = hay.chars().collect();
    let nd: Vec<char> = needle.chars().collect();
    for start in 0..=chars.len().saturating_sub(nd.len()) {
        if chars[start..start + nd.len()] != nd[..] {
            continue;
        }
        let before_ok = start == 0 || !is_word_char(chars[start - 1]);
        let after = start + nd.len();
        let after_ok = after >= chars.len() || !is_word_char(chars[after]);
        if before_ok && after_ok {
            return true;
        }
    }
    false
}
